### PR TITLE
First stage of cleanup in source tree pollution

### PR DESCRIPTION
**Pull request by sbc100**
Update build-js.sh to output to `out` directory.  This is district
from the `bin` directory which is used by the cmake build and may or
may not live in the source tree.  The `out` directory currently always
lives in the source tree.

As a followup change I hope to additionally move all test outout into
this tree.

See #2104

### CODE REVIEW

**Comment by sbc100** [comment]
I was a little confusing because BINARYEN_BIN is in the output directory (i.e. not necessarily under the source checkout at all).  But the build-js.sh always writes into the source directory.   

I'm using "out" as place where all kind of intermiates can be written, and "bin" as the place where the final binaries live.  In a followup I'm trying to have all tests write to the "out" directory too.

File: .gitignore [hunk 1]
```diff
@@ -3,14 +3,10 @@
 *.diff
 
 # autogenerated during the build
-src/passes/WasmIntrinsics.cpp
+/src/passes/WasmIntrinsics.cpp
```
Comment (dschuff) [inline .gitignore]: what is the effect of adding the leading `/` everywhere?

Comment (sbc100) [reply .gitignore]: IIUC is just makes it a more precise match.

e.g.
```
*.o
```
vs
```
/*.o
```
The latter only match objects at the top level

**Comment by kripken** [comment]
Sounds reasonable. I slightly worry because emcc does look at `$BINARYEN_ROOT/bin` for wasm-opt, wasm2js, etc. I guess we can update that if necessary?

**Comment by sbc100** [comment]
Those tools still build to "bin" and install to "bin".   But for out-of-tree build the "bin" the build to is $BUILDDIR/bin.

i'm not changing to locations of any of the cmake targets..  Only the build-js target

**Comment by kripken** [comment]
That should be fine then - I don't think we depend on the location of binaryen.js anymore anywhere.

**Pull request merged**